### PR TITLE
Add function to convert results to an xunit xml file

### PR DESCRIPTION
**Pull request by wheelerlaw**
Adds a function so that the test results are outputted as an xunit xml file that can be parsed by things that support it.

### CODE REVIEW

**Comment by lachmanfrantisek** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 10
           

Complexity increasing per file
==============================
- colin/cli/colin.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/user-cont/colin/pullRequest?prid=5390981&bid=17673051)

File: colin/core/result.py [hunk 1]
```diff
@@ -97,6 +99,44 @@ def save_json_to_file(self, file):
...
+                logs.append(log)
+
+        rough_string = ElementTree.toString(top, 'utf-8')
+        reparsed = minidom.parseString(rough_string)
+        return reparsed.toprettyxml(indent="  ")
+
+    def save_xunit_to_file(self, file):
+        f = open(file, 'w')
+        f.write(self.xunit)
+        f.close()
```
Comment (lachmanfrantisek) [inline colin/core/result.py:138]: You don't need to use `open`. The object you get from `click` is already writable.

**Comment by lachmanfrantisek** [comment]
[test]

**Comment by lslebodn** [comment]
@wheelerlaw would you mind to add simple  line to cli test? 
Similar as for json
https://github.com/user-cont/colin/blob/master/tests/unit/test_cli.py#L36

File: colin/core/result.py [hunk 2]
```diff
@@ -97,6 +99,44 @@ def save_json_to_file(self, file):
...
+            logs = SubElement(testcase, "logs")
+            for log in r.logs:
+                log = SubElement(logs, "log", {
+                    'message': log,
+                    'result': 'INFO',
+                    'waiver_authorization': 'Not Waivable'
+                })
+                logs.append(log)
+
+        rough_string = ElementTree.toString(top, 'utf-8')
```
Comment (lslebodn) [inline colin/core/result.py:131]: it is `tostring` and not `toString` :-)

BTW `tostring` is a function in the module xml.etree.elementtree
https://docs.python.org/3.6/library/xml.etree.elementtree.html#xml.etree.ElementTree.tostring

And here the class is used xml.etree.ElementTree.ElementTree which does not have such method.
https://docs.python.org/3.6/library/xml.etree.elementtree.html#xml.etree.ElementTree.ElementTree

File: colin/core/result.py [hunk 3]
```diff
@@ -97,6 +99,44 @@ def save_json_to_file(self, file):
...
+        """
+        Get the xunit representation of results
+
+        :return: str
+        """
+
+        top = Element("testsuites")
+
+        testsuite = SubElement(top, "testsuite")
+        top.append(testsuite)
```
Comment (lslebodn) [inline colin/core/result.py:113]: `SubElement(top, "testsuite")` already added child element to the top and calling append
will cause two the same elements in the `<testsuites>`

File: colin/core/result.py [hunk 4]
```diff
@@ -97,6 +99,44 @@ def save_json_to_file(self, file):
...
+
+        for r in self.results:
+            testcase = SubElement(testsuite, "testcase", {
+                'name': r.check_name,
+                'status': "PASSED" if r.ok else "FAILED", # Can't use PASSED or FAILED global variables because their values are PASS and FAIL respectively.
+                'url': r.reference_url
+            })
+            testsuite.append(testcase)
```
Comment (lslebodn) [inline colin/core/result.py:121]: The same case as for `SubElement(top, "testsuite")`

File: colin/core/result.py [hunk 5]
```diff
@@ -97,6 +99,44 @@ def save_json_to_file(self, file):
...
+            logs = SubElement(testcase, "logs")
+            for log in r.logs:
+                log = SubElement(logs, "log", {
+                    'message': log,
+                    'result': 'INFO',
+                    'waiver_authorization': 'Not Waivable'
+                })
+                logs.append(log)
```
Comment (lslebodn) [inline colin/core/result.py:129]: The same case as for `SubElement(top, "testsuite")`

**Comment by lachmanfrantisek** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 6
           

Complexity increasing per file
==============================
- colin/cli/colin.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/user-cont/colin/pullRequest?prid=5390981&bid=17673051)

File: colin/core/result.py (hunk 1 above)
Comment (wheelerlaw) [reply colin/core/result.py]: Thanks

**Comment by lachmanfrantisek** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 6
           

Complexity increasing per file
==============================
- colin/cli/colin.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/user-cont/colin/pullRequest?prid=5390981&bid=17673051)

**Comment by lslebodn** [comment]
@wheelerlaw could you rebase on top of current master.
I fixed some tests related issues in !238

**Comment by lachmanfrantisek** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 7
           

Complexity increasing per file
==============================
- colin/cli/colin.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/user-cont/colin/pullRequest?prid=5390981&bid=17673051)

**Comment by lslebodn** [comment]
[test]

**Comment by lslebodn** [comment]
ok to test

File: colin/core/result.py [hunk 6]
```diff
@@ -15,6 +15,8 @@
 #
 
 import json
+from xml.etree.ElementTree import Element, SubElement, ElementTree, tostring
```
Comment (lslebodn) [inline colin/core/result.py:18]: Codacy  now complains about unused  import
https://app.codacy.com/manual/user-cont/colin/pullRequest?prid=5390981

File: colin/cli/colin.py [hunk 7]
```diff
@@ -80,7 +82,7 @@ def cli():
               help="Timeout for each check in seconds. (default=600)")
 @click.option('--insecure', is_flag=True, default=False,
               help="Pull from an insecure registry (HTTP or invalid TLS).")
-def check(target, parent_target, ruleset, ruleset_file, debug, json, stat, skip, tag, verbose,
+def check(target, parent_target, ruleset, ruleset_file, debug, json, xunit, stat, skip, tag, verbose,
```
Comment (lslebodn) [inline colin/cli/colin.py:85]: Could you add verbose to next line?

https://app.codacy.com/manual/user-cont/colin/pullRequest?prid=5390981

File: colin/core/result.py [hunk 8]
```diff
@@ -97,6 +99,43 @@ def save_json_to_file(self, file):
...
+
+        for r in self.results:
+            testcase = SubElement(testsuite, "testcase", {
+                'name': r.check_name,
+                # Can't use PASSED or FAILED global variables because their values are PASS
+                # and FAIL respectively and xunit wants them suffixed with -ED.
+                'status': "PASSED" if r.ok else "FAILED",
+                'url': r.reference_url
+            })
+            if len(r.logs) > 0:
```
Comment (lslebodn) [inline colin/core/result.py:122]: It is enough to check `if r.logs` in python
https://app.codacy.com/manual/user-cont/colin/pullRequest?prid=5390981

File: tests/unit/test_cli.py [hunk 9]
```diff
@@ -34,6 +34,7 @@ def _common_help_options(result):
     assert "-f, --ruleset-file FILENAME" in result.output
     assert "--debug" in result.output
     assert "--json FILENAME" in result.output
+    assert "--xunit FILENAME" in result.output
```
Comment (lslebodn) [inline tests/unit/test_cli.py:37]: Sorry for confusion but `_common_help_options` is used on more places
and we added that option just to `colin check`.

Therefore test `test_list_checks_help_command` failed
 
https://ci.centos.org/job/user-cont-colin-pr/117/

Moving the assertion to the `test_check_help_command` should fix that

**Comment by lslebodn** [comment]
@wheelerlaw I think it is not WIP anymore.

We get xunit output and we can modify/add attributes/elements later after seeing results in webui. But we need to integrate at least something to the dashboard :-)

File: colin/core/result.py [hunk 10]
```diff
@@ -97,6 +99,43 @@ def save_json_to_file(self, file):
...
+
+        testsuite = SubElement(top, "testsuite")
+
+        for r in self.results:
+            testcase = SubElement(testsuite, "testcase", {
+                'name': r.check_name,
+                # Can't use PASSED or FAILED global variables because their values are PASS
+                # and FAIL respectively and xunit wants them suffixed with -ED.
+                'status': "PASSED" if r.ok else "FAILED",
+                'url': r.reference_url
```
Comment (wheelerlaw) [inline colin/core/result.py:120]: So everything appears to be working, but the only issue I am having is what other fields to put here? The JSON output has more fields (name, ok, status, description, message, reference_url, logs), however some of the example xunit files you sent me don't have those fields. So where would I put those things?

**Comment by wheelerlaw** [comment]
> @wheelerlaw I think it is not WIP anymore.
> 
> We get xunit output and we can modify/add attributes/elements later after seeing results in webui. But we need to integrate at least something to the dashboard :-)

Okay, I removed the WIP designator.

**Comment by lachmanfrantisek** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- colin/cli/colin.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/user-cont/colin/pullRequest?prid=5390981&bid=17673051)

File: colin/core/result.py (hunk 10 above)
Comment (lslebodn) [reply colin/core/result.py:120]: We can modify it later.
So far LGTM

**Comment by lslebodn** [comment]
@lachmanfrantisek Do you have an idea why 2 packit tests has not finished yet?
Is it a packit bug?

**Comment by lachmanfrantisek** [comment]
Looks like some problem in the testing-farm:

```
[21:11:26] [E] Pipeline reported an exception in module 'cruncher': Failed to boot test machine with image '/opt/cruncher/images/Fedora-Cloud-Base-30-20200423.0.x86_64.qcow2', is /dev/kvm available? See 'provision' logs for more details.[21:11:26] [E] Exiting with status -1
```

@lslebodn @wheelerlaw thanks for this enhancement.

**Pull request merged**